src/simulator/ButtonModule.tsx

```diff
@@ -50,81 +50,86 @@ const SensorButton = ({
-            handleSensorChange(sensor.max);
-          } else {
-            handleSensorChange(sensor.min);
-          }
-      }
-    };
-    const mouseDownListener = (event: MouseEvent) => {
-      event.preventDefault();
-      handleSensorChange(sensor.max);
-    };
-    const mouseUpListener = (event: MouseEvent) => {
-      event.preventDefault();
-      handleSensorChange(sensor.min);
-    };
-    const mouseLeaveListener = () => {
+  const keyListener = (event: React.KeyboardEvent<HTMLButtonElement>) => {
+    switch (event.key) {
+      case "Enter":
+      case " ":
+        event.preventDefault();
+        if (event.type === "keydown") {
+          handleSensorChange(sensor.max);
+        } else {
+          handleSensorChange(sensor.min);
+        }
+    }
+  };
+  const mouseDownListener = (
+    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
+  ) => {
+    event.preventDefault();
+    setMouseDown(true);
+    handleSensorChange(sensor.max);
+  };
+  const mouseUpListener = (
+    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
+  ) => {
+    event.preventDefault();
+    setMouseDown(false);
+    handleSensorChange(sensor.min);
+  };
+  const mouseLeaveListener = () => {
+    if (mouseDown) {
       handleSensorChange(sensor.min);
-    };
-    if (buttonEl) {
-      buttonEl.addEventListener("mousedown", mouseDownListener);
-      buttonEl.addEventListener("mouseup", mouseUpListener);
-      buttonEl.addEventListener("keydown", keyListener);
-      buttonEl.addEventListener("keyup", keyListener);
-      buttonEl.addEventListener("mouseleave", mouseLeaveListener);
     }
-    return () => {
-      if (buttonEl) {
-        buttonEl.removeEventListener("mousedown", mouseDownListener);
-        buttonEl.removeEventListener("mouseup", mouseUpListener);
-        buttonEl.removeEventListener("keydown", keyListener);
-        buttonEl.removeEventListener("keyup", keyListener);
-        buttonEl.removeEventListener("mouseleave", mouseLeaveListener);
-      }
-    };
-  }, [handleSensorChange, sensor]);
-  const handleOverrideSet = (event: React.ChangeEvent<HTMLInputElement>) => {
-    setIsPressedOverride(event.currentTarget.checked);
-    handleSensorChange(
-      event.currentTarget.checked ? sensor.max : sensor.min,
-      true
-    );
   };
-  const disabled = simState === SimState.STOPPED;
+  const [isHeld, setIsHeld] = useState<boolean>(false);
+  const handleOverrideSet = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setIsHeld(event.currentTarget.checked);
+      handleSensorChange(event.currentTarget.checked ? sensor.max : sensor.min);
+    },
+    [handleSensorChange, setIsHeld, sensor]
+  );
+
   useEffect(() => {
-    if (disabled) {
-      setIsPressedOverride(false);
+    if (sensor.value === sensor.min) {
+      setIsHeld(false);
+    }
+    if (sensor.value === sensor.max) {
+      setIsHeld(true);
     }
-  }, [disabled]);
+  }, [sensor]);
```

Comment:
Can this be done in an event handler?

Reply:
I think we need the effect because we listen to sensor changes resulting from interaction with the SVG board as well as the in Editor buttons/switches?

Reply:
Ah, I see what you mean. Discussing a possible alternative on chat but fine with it like this.

Reply:
Contrary to what I said in chat earlier, this appears to be better to the `useEffect` alternative -> [`720cfb6`](https://github.com/microbit-foundation/python-editor-next/pull/882/commits/720cfb67826e97250951cb97ed0d6498295d7344).

Reply:
I guess the alternative is to pull the button state up so it can be reconciled with the sensor state at the point where it changes (+ some change to the model to store whether it's held). This is fine though as it keeps it local to buttons.

Thanks.